// @aigne/agent-library/lib/esm/orchestrator/orchestrator-prompts.js
import { ensureZodUnionArray } from "@aigne/core/utils/json-schema.js";
import { duplicates } from "@aigne/core/utils/type-utils.js";
import { z } from "zod";
/**
 * @hidden
 */
export const SYNTHESIZE_PLAN_USER_PROMPT_TEMPLATE = `\
Synthesize the results of executing all steps in the plan into a cohesive result
`;
/**
 * @hidden
 */
export function getFullPlanSchema(agents) {
    const agentNames = agents.map((i) => i.name);
    if (new Set(agentNames).size !== agentNames.length) {
        const dup = duplicates(agentNames);
        throw new Error(`Duplicate agent names found in orchestrator: ${dup.join(",")}`);
    }
    const TaskSchema = z.object({
        description: z.string().describe("Detailed description of the task"),
        agent: z
            .union(ensureZodUnionArray(agents.map((i) => z.literal(i.name))))
            .describe("Name of the agent to execute the task"),
    });
    const StepSchema = z.object({
        description: z.string().describe("Detailed description of the step"),
        tasks: z.array(TaskSchema).describe("Tasks that can run in parallel in this step"),
    });
    return z.object({
        steps: z.array(StepSchema).describe("All sequential steps in the plan"),
        isComplete: z.boolean().describe("Whether the previous plan results achieve the objective"),
    });
}
/**
 * @hidden
 */
export const FULL_PLAN_PROMPT_TEMPLATE = `You are tasked with orchestrating a plan to complete an objective.
You can analyze results from the previous steps already executed to decide if the objective is complete.
Your plan must be structured in sequential steps, with each step containing independent parallel subtasks.

<objective>
{{objective}}
</objective>

<steps_completed>
{% for step in steps %}
- Step: {{step.description}}
  Result: {{result}}
{% endfor %}
</steps_completed>

<previous_plan_status>
{{plan.status}}
</previous_plan_status>

<previous_plan_result>
{{plan.result}}
</previous_plan_result>

You have access to the following Agents(which are collections of tools/functions):

<agents>
{% for agent in agents %}
- Agent: {{agent.name}}
  Description: {{agent.description}}
  Functions:
    {% for tool in tools %}
    - Tool: {{tool.name}}
      Description: {{tool.description}}
    {% endfor %}
{% endfor %}
</agents>

- If the previous plan results achieve the objective, return isComplete=true.
- Otherwise, generate remaining steps needed.
- Generate a plan with all remaining steps needed.
- Steps are sequential, but each Step can have parallel subtasks.
- For each Step, specify a description of the step and independent subtasks that can run in parallel.
- For each subtask specify:
    1. Clear description of the task that an LLM can execute
    2. Name of 1 Agent to use for the task`;
/**
 * @hidden
 */
export const TASK_PROMPT_TEMPLATE = `\
You are part of a larger workflow to achieve the step then the objective:

<objective>
{{objective}}
</objective>

<step>
{{step.description}}
</step>

Your job is to accomplish only the following task:

<task>
{{task.description}}
</task>

Results so far that may provide helpful context:

<steps_completed>
{% for step in steps %}
- Step: {{step.description}}
  Result: {{step.result}}
{% endfor %}
</steps_completed>
`;
/**
 * @hidden
 */
export const SYNTHESIZE_STEP_PROMPT_TEMPLATE = `\
Synthesize the results of these parallel tasks into a cohesive result

<objective>
{{objective}}
</objective>

<step>
{{step.description}}
</step>

<tasks>
{% for task in tasks %}
- Task: {{task.description}}
  Result: {{task.result}}
{% endfor %}
</tasks>
`;
