// @aigne/ideogram/lib/esm/ideogram-image-model.js
import { ImageModel, imageModelInputSchema, } from "@aigne/core";
import { snakelize } from "@aigne/core/utils/camelize.js";
import { checkArguments, pick } from "@aigne/core/utils/type-utils.js";
import { joinURL } from "ufo";
import { z } from "zod";
const IDEOGRAM_BASE_URL = "https://api.ideogram.ai";
const ideogramImageModelInputSchema = imageModelInputSchema.extend({});
const ideogramImageModelOptionsSchema = z.object({
    apiKey: z.string().optional(),
    baseURL: z.string().optional(),
    model: z.string().optional(),
    modelOptions: z.object({}).optional(),
});
export class IdeogramImageModel extends ImageModel {
    options;
    constructor(options) {
        super({
            ...options,
            inputSchema: ideogramImageModelInputSchema,
            description: options?.description ?? "Draw or edit image by Ideogram image models",
        });
        this.options = options;
        if (options)
            checkArguments(this.name, ideogramImageModelOptionsSchema, options);
    }
    apiKeyEnvName = "IDEOGRAM_API_KEY";
    get credential() {
        return {
            url: this.options?.baseURL || process.env.IDEOGRAM_BASE_URL || IDEOGRAM_BASE_URL,
            apiKey: this.options?.apiKey || process.env[this.apiKeyEnvName],
        };
    }
    get modelOptions() {
        return this.options?.modelOptions;
    }
    /**
     * Process the input and generate a response
     * @param input The input to process
     * @returns The generated response
     */
    async process(input) {
        const model = input.model;
        const formData = new FormData();
        if (model !== "ideogram-v3") {
            throw new Error(`${this.name} only support ideogram-v3`);
        }
        const inputKeys = [
            "prompt",
            "seed",
            "resolution",
            "aspectRatio",
            "renderingSpeed",
            "magicPrompt",
            "negativePrompt",
            "colorPalette",
            "styleCodes",
            "styleType",
        ];
        const mergedInput = snakelize(pick({ ...this.modelOptions, ...input }, inputKeys));
        if (input.n) {
            formData.append("num_images", input.n.toString());
        }
        Object.keys(mergedInput).forEach((key) => {
            if (mergedInput[key]) {
                formData.append(key, mergedInput[key]);
            }
        });
        const { url, apiKey } = this.credential;
        if (!apiKey)
            throw new Error(`${this.name} requires an API key. Please provide it via \`options.apiKey\`, or set the \`${this.apiKeyEnvName}\` environment variable`);
        const response = await fetch(joinURL(new URL(url).origin, `/v1/${model}/generate`), {
            method: "POST",
            headers: { "api-key": apiKey },
            body: formData,
        });
        if (!response.ok) {
            const error = await response.text();
            throw new Error(`Ideogram API error: ${response.status} ${response.statusText} ${error}`);
        }
        const data = await response.json();
        return {
            images: data.data.map((item) => ({ url: item.url })),
            usage: {
                inputTokens: 0,
                outputTokens: 0,
            },
            model,
        };
    }
}
