// @aigne/core/lib/cjs/utils/role-utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.mapStandardRole = exports.STANDARD_ROLE_MAP = void 0;
exports.createRoleMapper = createRoleMapper;
/**
 * Standard role mapping for most chat model providers
 * Maps AIGNE framework roles to common provider role names
 */
exports.STANDARD_ROLE_MAP = {
    system: "system",
    user: "user",
    agent: "assistant",
    tool: "tool",
};
/**
 * Creates a role mapper function for a specific provider
 *
 * @param roleMap - Custom role mapping for the provider
 * @returns Function that maps AIGNE roles to provider roles
 *
 * @example
 * ```typescript
 * // For standard providers (OpenAI, Anthropic, etc.)
 * const mapRole = createRoleMapper(STANDARD_ROLE_MAP);
 *
 * // For providers with different role names
 * const customMap = { ...STANDARD_ROLE_MAP, agent: "bot" };
 * const customMapper = createRoleMapper(customMap);
 * ```
 */
function createRoleMapper(roleMap) {
    return (role) => roleMap[role];
}
/**
 * Standard role mapper using the default role mapping
 */
exports.mapStandardRole = createRoleMapper(exports.STANDARD_ROLE_MAP);
