// @aigne/core/lib/cjs/utils/type-utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isNil = isNil;
exports.isRecord = isRecord;
exports.isEmpty = isEmpty;
exports.isNonNullable = isNonNullable;
exports.isNotEmpty = isNotEmpty;
exports.duplicates = duplicates;
exports.remove = remove;
exports.unique = unique;
exports.pick = pick;
exports.omit = omit;
exports.omitDeep = omitDeep;
exports.omitBy = omitBy;
exports.flat = flat;
exports.createAccessorArray = createAccessorArray;
exports.checkArguments = checkArguments;
exports.tryOrThrow = tryOrThrow;
const zod_1 = require("zod");
function isNil(value) {
    return value === null || value === undefined;
}
function isRecord(value) {
    return !!value && typeof value === "object" && !Array.isArray(value);
}
function isEmpty(obj) {
    if (isNil(obj))
        return true;
    if (typeof obj === "string" || Array.isArray(obj))
        return obj.length === 0;
    if (typeof obj === "object")
        return Object.keys(obj).length === 0;
    return false;
}
function isNonNullable(value) {
    return !isNil(value);
}
function isNotEmpty(arr) {
    return arr.length > 0;
}
function duplicates(arr, key = (item) => item) {
    const seen = new Set();
    const duplicates = new Set();
    for (const item of arr) {
        const k = key(item);
        if (seen.has(k)) {
            duplicates.add(item);
        }
        else {
            seen.add(k);
        }
    }
    return Array.from(duplicates);
}
function remove(arr, remove) {
    const removed = [];
    for (let i = 0; i < arr.length; i++) {
        // biome-ignore lint/style/noNonNullAssertion: false positive
        const item = arr[i];
        if ((Array.isArray(remove) && remove.includes(item)) ||
            (typeof remove === "function" && remove(item))) {
            removed.push(...arr.splice(i, 1));
            i--;
        }
    }
    return removed;
}
function unique(arr, key = (item) => item) {
    const seen = new Set();
    return arr.filter((item) => {
        const k = key(item);
        if (seen.has(k)) {
            return false;
        }
        seen.add(k);
        return true;
    });
}
function pick(obj, ...keys) {
    const flattenedKeys = new Set(keys.flat());
    return Object.fromEntries(Object.entries(obj).filter(([key]) => flattenedKeys.has(key)));
}
function omit(obj, ...keys) {
    const flattenedKeys = new Set(keys.flat());
    return Object.fromEntries(Object.entries(obj).filter(([key]) => !flattenedKeys.has(key)));
}
function omitDeep(obj, ...keys) {
    if (Array.isArray(obj)) {
        return obj.map((item) => omitDeep(item, ...keys));
    }
    if (isRecord(obj)) {
        const flattenedKeys = new Set(keys.flat());
        return Object.fromEntries(Object.entries(obj)
            .filter(([key]) => !flattenedKeys.has(key))
            .map(([key, value]) => [key, omitDeep(value, ...keys)]));
    }
    return obj;
}
function omitBy(obj, predicate) {
    return Object.fromEntries(Object.entries(obj).filter(([key, value]) => {
        const k = key;
        return !predicate(value, k);
    }));
}
function flat(...value) {
    return value.flat().filter(isNonNullable);
}
function createAccessorArray(array, accessor) {
    return new Proxy(array, {
        get: (t, p, r) => Reflect.get(t, p, r) ?? accessor(array, p),
    });
}
function checkArguments(prefix, schema, args) {
    try {
        return schema.parse(args, {
            errorMap: (issue, ctx) => {
                if (issue.code === "invalid_union") {
                    // handle all issues that are not invalid_type
                    const otherQuestions = issue.unionErrors
                        .map(({ issues: [issue] }) => {
                        if (issue && issue.code !== "invalid_type") {
                            return issue.message || zod_1.z.defaultErrorMap(issue, ctx).message;
                        }
                    })
                        .filter(isNonNullable);
                    if (otherQuestions.length) {
                        return { message: otherQuestions.join(", ") };
                    }
                    // handle invalid_type issues
                    const expected = issue.unionErrors
                        .map(({ issues: [issue] }) => {
                        if (issue?.code === "invalid_type") {
                            return issue;
                        }
                    })
                        .filter(isNonNullable);
                    if (expected.length) {
                        return {
                            message: `Expected ${expected.map((i) => i.expected).join(" or ")}, received ${expected[0]?.received}`,
                        };
                    }
                }
                return { message: ctx.defaultError };
            },
        });
    }
    catch (error) {
        if (error instanceof zod_1.z.ZodError) {
            const message = error.issues.map((i) => `${i.path}: ${i.message}`).join(", ");
            throw new Error(`${prefix} check arguments error: ${message}`);
        }
        throw error;
    }
}
function tryOrThrow(fn, error) {
    const createError = (e) => {
        return typeof error === "function"
            ? error(e)
            : typeof error === "string"
                ? new Error(error)
                : error;
    };
    try {
        const result = fn();
        if (result instanceof Promise) {
            return result.catch((e) => {
                const error = createError(e);
                if (error)
                    throw error;
            });
        }
        return result;
    }
    catch (e) {
        const error = createError(e);
        if (error)
            throw error;
    }
}
