// @aigne/platform-helpers/lib/cjs/nodejs/index.node.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.nodejs = void 0;
const node_crypto_1 = __importDefault(require("node:crypto"));
const node_fs_1 = __importDefault(require("node:fs"));
const promises_1 = __importDefault(require("node:fs/promises"));
const node_os_1 = __importDefault(require("node:os"));
const node_path_1 = __importDefault(require("node:path"));
const node_util_1 = require("node:util");
exports.nodejs = {
    customInspect: node_util_1.inspect.custom,
    get isStdoutATTY() {
        return process.stdout?.isTTY;
    },
    get isStderrATTY() {
        return process.stderr?.isTTY;
    },
    get env() {
        return process.env;
    },
    get fs() {
        return promises_1.default;
    },
    get fsSync() {
        return node_fs_1.default;
    },
    get path() {
        return node_path_1.default;
    },
    get os() {
        return node_os_1.default;
    },
    get crypto() {
        return node_crypto_1.default;
    },
};
