// @aigne/core/lib/esm/memory/retriever.js
import { z } from "zod";
import { Agent, } from "../agents/agent.js";
/**
 * @hidden
 */
export const memoryRetrieverInputSchema = z.object({
    limit: z.number().optional(),
    search: z.union([z.string(), z.record(z.string(), z.unknown())]).optional(),
});
/**
 * @hidden
 */
export const memoryRetrieverOutputSchema = z.object({
    memories: z.array(z.object({
        id: z.string(),
        content: z.custom(),
        createdAt: z.string().datetime(),
    })),
});
/**
 * Abstract base class for agents that retrieve memories from storage.
 *
 * The MemoryRetriever serves as a foundation for implementing specific memory
 * retrieval mechanisms. Implementations of this class are responsible for:
 *
 * 1. Querying a memory storage backend to find relevant memories
 * 2. Filtering memories based on search criteria
 * 3. Limiting the number of results returned
 * 4. Potentially implementing sorting, ranking, or relevance-based retrieval
 *
 * Custom implementations should extend this class and provide concrete
 * implementations of the process method to handle the actual retrieval logic.
 */
export class MemoryRetriever extends Agent {
    tag = "MemoryRetrieverAgent";
    /**
     * Creates a new MemoryRetriever instance with predefined input and output schemas.
     *
     * @param options - Configuration options for the memory retriever agent
     */
    constructor(options) {
        super({
            ...options,
            inputSchema: memoryRetrieverInputSchema,
            outputSchema: memoryRetrieverOutputSchema,
        });
        this._process = options.process;
    }
    _process;
    process(input, options) {
        if (!this._process) {
            throw new Error("MemoryRetriever process function is not implemented.");
        }
        return this._process(input, options);
    }
}
