// @aigne/transport/lib/cjs/http-client/client-agent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClientAgent = void 0;
const core_1 = require("@aigne/core");
const stream_utils_js_1 = require("@aigne/core/utils/stream-utils.js");
class ClientAgent extends core_1.Agent {
    client;
    tag = "ClientAgent";
    constructor(client, options) {
        super(options);
        this.client = client;
    }
    async invoke(input, options) {
        const memories = await this.retrieveMemories({ search: input }, { ...options, context: this.client });
        const result = await this.client._invoke(this.name, input, {
            ...options,
            returnActiveAgent: false,
            memories,
        });
        if (!(result instanceof ReadableStream)) {
            await this.postprocess(input, result, {
                ...options,
                context: this.client,
            });
            return result;
        }
        return (0, stream_utils_js_1.onAgentResponseStreamEnd)(result, {
            onResult: async (result) => {
                await this.postprocess(input, result, {
                    ...options,
                    context: this.client,
                });
            },
        });
    }
    process(_input, _options) {
        throw new Error("Method not implemented.");
    }
}
exports.ClientAgent = ClientAgent;
