// @aigne/agentic-memory/lib/cjs/memory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AgenticMemoryRecorder = exports.AgenticMemoryRetriever = exports.AgenticMemory = void 0;
const core_1 = require("@aigne/core");
const type_utils_js_1 = require("@aigne/core/utils/type-utils.js");
const default_memory_1 = require("@aigne/default-memory");
const zod_1 = require("zod");
const prompt_js_1 = require("./prompt.js");
class AgenticMemory extends core_1.MemoryAgent {
    constructor(options = {}) {
        const storage = options.storage instanceof default_memory_1.MemoryStorage
            ? options.storage
            : new default_memory_1.DefaultMemoryStorage(options.storage);
        super({
            ...options,
            recorder: options.recorder ?? new AgenticMemoryRecorder({ ...options, storage }),
            retriever: options.retriever ?? new AgenticMemoryRetriever({ ...options, storage }),
            autoUpdate: options.autoUpdate ?? true,
        });
        this.storage = storage;
    }
    storage;
}
exports.AgenticMemory = AgenticMemory;
class AgenticMemoryRetriever extends default_memory_1.DefaultMemoryRetriever {
}
exports.AgenticMemoryRetriever = AgenticMemoryRetriever;
class AgenticMemoryRecorder extends core_1.MemoryRecorder {
    constructor(options) {
        super(options);
        this.storage = options.storage;
        this.inputKey = (0, type_utils_js_1.flat)(options.inputKey);
        this.outputKey = (0, type_utils_js_1.flat)(options.outputKey);
        this.agent =
            options.agent ??
                core_1.AIAgent.from({
                    name: "agentic_memory_extractor",
                    description: "Records memories in files by AI agent",
                    instructions: options.instructions || prompt_js_1.DEFAULT_AGENTIC_MEMORY_RECORDER_INSTRUCTIONS,
                    outputSchema: zod_1.z.object({
                        newMemories: zod_1.z
                            .array(zod_1.z.object({
                            content: zod_1.z.string().describe("Content of the memory"),
                        }))
                            .describe("Newly created memories"),
                    }),
                });
    }
    storage;
    inputKey;
    outputKey;
    agent;
    async process(input, options) {
        const agenticMemories = await options.context.invoke(this.agent, {
            content: input.content.map((item) => ({
                input: item.input && this.inputKey?.length ? (0, type_utils_js_1.pick)(item.input, this.inputKey) : item.input,
                output: item.output && this.outputKey?.length ? (0, type_utils_js_1.pick)(item.output, this.outputKey) : item.output,
                source: item.source,
            })),
        });
        const newMemories = [];
        for (const item of agenticMemories.newMemories) {
            const { result } = await this.storage.create({ content: item.content }, options);
            newMemories.push(result);
        }
        return {
            memories: newMemories,
        };
    }
}
exports.AgenticMemoryRecorder = AgenticMemoryRecorder;
