// @aigne/core/lib/esm/utils/logger.js
import { nodejs } from "@aigne/platform-helpers/nodejs/index.js";
import debug from "debug";
export var LogLevel;
(function (LogLevel) {
    LogLevel["SILENT"] = "silent";
    LogLevel["ERROR"] = "error";
    LogLevel["WARN"] = "warn";
    LogLevel["INFO"] = "info";
    LogLevel["DEBUG"] = "debug";
})(LogLevel || (LogLevel = {}));
const levels = Object.values(LogLevel);
export function getLevelFromEnv(ns) {
    for (const level of [...levels].reverse()) {
        if (debug.enabled(`${ns}:${level}`)) {
            return level;
        }
    }
}
export class Logger {
    options;
    constructor(options) {
        this.options = options;
        this.debugLogger = debug(`${options.ns}:${LogLevel.DEBUG}`);
        this.infoLogger = debug(`${options.ns}:${LogLevel.INFO}`);
        this.warnLogger = debug(`${options.ns}:${LogLevel.WARN}`);
        this.errorLogger = debug(`${options.ns}:${LogLevel.ERROR}`);
        this.level = getLevelFromEnv(options.ns) || options.level;
        for (const logger of [this.debugLogger, this.infoLogger, this.warnLogger]) {
            // @ts-ignore
            logger.useColors = nodejs.isStdoutATTY;
            logger.enabled = true;
            logger.log = (...args) => this.logMessage(...args);
        }
        this.errorLogger.log = (...args) => this.logError(...args);
        // @ts-ignore
        this.errorLogger.useColors = nodejs.isStderrATTY;
        this.errorLogger.enabled = true;
    }
    level;
    debugLogger;
    infoLogger;
    warnLogger;
    errorLogger;
    logMessage = console.log;
    logError = console.error;
    debug(message, ...args) {
        if (this.enabled(LogLevel.DEBUG)) {
            this.debugLogger(message, ...args);
        }
    }
    info(message, ...args) {
        if (this.enabled(LogLevel.INFO)) {
            this.infoLogger(message, ...args);
        }
    }
    warn(message, ...args) {
        if (this.enabled(LogLevel.WARN)) {
            this.warnLogger(message, ...args);
        }
    }
    error(message, ...args) {
        if (this.enabled(LogLevel.ERROR)) {
            this.errorLogger(message, ...args);
        }
    }
    enabled(level) {
        return levels.indexOf(this.level) >= levels.indexOf(level);
    }
}
export const logger = new Logger({
    ns: "aigne:core",
    level: LogLevel.SILENT,
});
