// @aigne/transport/lib/esm/http-client/client-agent.js
import { Agent, } from "@aigne/core";
import { onAgentResponseStreamEnd } from "@aigne/core/utils/stream-utils.js";
export class ClientAgent extends Agent {
    client;
    tag = "ClientAgent";
    constructor(client, options) {
        super(options);
        this.client = client;
    }
    async invoke(input, options) {
        const memories = await this.retrieveMemories({ search: input }, { ...options, context: this.client });
        const result = await this.client._invoke(this.name, input, {
            ...options,
            returnActiveAgent: false,
            memories,
        });
        if (!(result instanceof ReadableStream)) {
            await this.postprocess(input, result, {
                ...options,
                context: this.client,
            });
            return result;
        }
        return onAgentResponseStreamEnd(result, {
            onResult: async (result) => {
                await this.postprocess(input, result, {
                    ...options,
                    context: this.client,
                });
            },
        });
    }
    process(_input, _options) {
        throw new Error("Method not implemented.");
    }
}
